src/librustc_mir/util/borrowck_errors.rs

```diff
@@ -630,6 +630,28 @@ pub trait BorrowckErrors<'cx>: Sized + Copy {
         self.cancel_if_wrong_origin(err, o)
     }
 
+    fn cannot_return_reference_to_local(
+        self,
+        span: Span,
+        reference_desc: &str,
+        path_desc: &str,
+        o: Origin,
+    ) -> DiagnosticBuilder<'cx> {
+        let mut err = struct_span_err!(
+            self,
+            span,
+            E0515,
+            "cannot return {REFERENCE} {LOCAL}{OGN}",
+            REFERENCE=reference_desc,
+            LOCAL=path_desc,
+            OGN = o
+        );
+
+        err.span_label(span, "cannot return this value");
```

Comment:
I love the idea of this error, but I find the phrasing a bit terse (as I noted elsewhere)

Reply:
Looking at this example:

```
error[E0515]: cannot return reference to function parameter `x`
  --> $DIR/borrowck-local-borrow-outlives-fn.rs:15:5
   |
LL |     &x
   |     ^^ cannot return this value
```

Nowhere here does it took about *why* I can't do this. I think I would prefer something like this:

```
error[E0515]: cannot return reference to function parameter `x`
  --> $DIR/borrowck-local-borrow-outlives-fn.rs:15:5
   |
LL |     &x
   |     ^^ returns a reference to data owned by the current function
```